src/Contracts/TerminalService.ts

```diff
@@ -0,0 +1,86 @@
+import WebSocket = require('isomorphic-ws');
+import { Terminal } from 'xterm';
+import * as fit from 'xterm/lib/addons/fit/fit';
+Terminal.applyAddon(fit);
+export class TerminalHandler {
+    private server:string;
+    private token:string;
+    private termRef: HTMLElement | null;
+    private terminal:Terminal | null = null;
+    private socket:WebSocket | null = null;
+    constructor(config:any, termRef: HTMLDivElement | null) {
+        const ssl = config.server.startsWith('https://');
+        const target  = ssl ? config.server.substr(8) : config.server.substr(7);
+        const proto = ssl ? 'wss' : 'ws';
+        this.server = `${proto}://${target}`;
+        const b64enctoken = btoa(config.token);
+        this.token = b64enctoken.substring(0, b64enctoken.indexOf('=')); //stripping the = padding in base64 encoded token
+        this.termRef = termRef;
+    }
+
+    public generateTTYForPod(podName: string, containerName:string):void {
+        const protocols = ['base64url.bearer.authorization.k8s.io.','channel.k8s.io',];
+        //todo: put in a fallback if bin/bash is not found. Should be trying for sh, zsh, csh and powershell for windows containers
+        const path = '/api/v1/namespaces/default/pods/'+podName+'/exec?stdin=1&stdout=1&stderr=1&tty=1&command='+encodeURIComponent("/bin/bash")+'&container='+containerName;
+        const uri = `${this.server}${path}`;
+        //todo: service account already provides token in base64encoded format. Account for the token from service account
+        protocols[0] = protocols[0]+this.token;
+        const sock = new WebSocket(uri, protocols);
+        sock.onopen = () => {
+            this.socket = sock;
+            if(this.termRef) {
+                this.terminal = new Terminal({
+                    cols:180,
+                    rows:30,
+                    fontSize:14,
+                    cursorBlink:true,
+                });
+                this.terminal.on('data', (data)=>{
+                    this.transferData(data);
+                });
+                this.terminal.open(this.termRef);
+                (this.terminal as any).fit();
+                this.terminal.focus();
+            }
+        };
+
+        sock.onerror = (err) =>{
+            console.log(err);
+            this.socket = null;
```

Comment:
common code used in more than 2 places

this.socket = null;
            this.terminal && this.terminal.dispose();
            this.terminal = null;